theano/tensor/nnet/nnet.py

```diff
@@ -130,329 +132,486 @@ def c_code_template(dtype):
+            // Compute the stride on the selected dimension
+            stride_dx = PyArray_STRIDE(%(dx)s, axis)/sizeof(dtype_%(dx)s);;
+            stride_sm = PyArray_STRIDE(%(sm)s, axis)/sizeof(dtype_%(sm)s);
+            stride_dy = PyArray_STRIDE(%(dy)s, axis)/sizeof(dtype_%(dy)s);
+            // Get the shape on the specified dimension
+            num_sm_classes = shape_sm[axis];
+            // Go through the array
+            dtype_%(dx)s *dx_i;
+            dtype_%(sm)s *sm_i;
+            dtype_%(dy)s *dy_i;
+            while(PyArray_ITER_NOTDONE(it_dx))
+            {
+                dx_i = (dtype_%(dx)s *) PyArray_ITER_DATA(it_dx);
+                sm_i = (dtype_%(sm)s *) PyArray_ITER_DATA(it_sm);
+                dy_i = (dtype_%(dy)s *) PyArray_ITER_DATA(it_dy);
+
+                double sum_dy_times_sm = 0.;
+                for (size_t j = 0; j < num_sm_classes; ++j)
+                {
+                    dx_i[j] = dy_i[j] * sm_i[j];
+                    sum_dy_times_sm += dx_i[j];
+                }
+                for (size_t j = 0; j < num_sm_classes; ++j)
+                {
+                    dx_i[j] -= sum_dy_times_sm * sm_i[j];
+                }
+                PyArray_ITER_NEXT(it_dx);
+                PyArray_ITER_NEXT(it_sm);
+                PyArray_ITER_NEXT(it_dy);
+            }
+            ''' % dict(locals(), **sub)
+
 softmax_grad = SoftmaxGrad()
 
 
-class Softmax(gof.Op):
+class Abstract_softmax(gof.Op):
     """
     Softmax activation function
     :math:`\\varphi(\\mathbf{x})_j =
     \\frac{e^{\mathbf{x}_j}}{\sum_{k=1}^K e^{\mathbf{x}_k}}`
     where :math:`K` is the total number of neurons in the layer. This
-    activation function gets applied row-wise.
-
+    activation function gets applied on the selected axes (Default axis is the last).
     """
-
-    nin = 1
+    nin = 2  # tensor, axis
     nout = 1
+    E_axis = 'invalid axis'
     __props__ = ()
 
-    def make_node(self, x):
+    def make_node(self, x, axis=-1):
         x = tensor.as_tensor_variable(x)
-        if x.type.ndim not in (1, 2) \
-                or x.type.dtype not in tensor.float_dtypes:
-            raise ValueError('x must be 1-d or 2-d tensor of floats. Got %s' %
-                             x.type)
+        axis = tensor.as_tensor_variable(axis)
+        # TODO : Delete this and modify the test accordly
         if x.ndim == 1:
             x = tensor.shape_padleft(x, n_ones=1)
-
-        return Apply(self, [x], [x.type()])
+        inputs = [x, axis]
+        return Apply(self, inputs, [x.type()])
 
     def perform(self, node, input_storage, output_storage):
-        x, = input_storage
-        e_x = np.exp(x - x.max(axis=1)[:, None])
-        sm = e_x / e_x.sum(axis=1)[:, None]
+        x, axes = input_storage
+        # Apply softmax on the specified dimension
+        e_x = np.exp(x - x.max(axis=axes, keepdims=True))
```

Comment:
Try this example:
```python
a = T.tensor3('a')
axis = T.constant([0, 1, -1])
res = softmax(a, axis=axis)
f = theano.function([a], res)
a_ = np.random.rand(3, 4, 5)
res_ = f(a_)
print res_.shape
```
When we reach here, ```axes``` is ```array([0, 1, 2])``` which leads to an error. A fix that comes to my mind is to cast it to tuple like ```x.max(axis=tuple(axes), keepdims=True)```.